Bug: Source code parsing fails on identical function prototypes in separate files.

## Describe the bug

If you have e.g. "int main(void)" in two files that are parsed with the  REQUIREMENT_TO_SOURCE_TRACEABILITY feature it will crash.

## To Reproduce


## Expected behavior

Would be nice to not have to exclude all files with a main(). Sometimes there is also the case where two different .c files implement the same functions from a common header, e.g. Hardware Abstraction Layers for different platforms.

## Screenshots

```
Traceback (most recent call last):
  File "/home/johan/virtual_python/bin/strictdoc", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 194, in main
    raise exc
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 189, in main
    _main(parallelizer)
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 74, in _main
    export_action.build_index()
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/helpers/timing.py", line 18, in wrap
    result = func(*args, **kw)
             ^^^^^^^^^^^^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/core/actions/export_action.py", line 42, in build_index
    TraceabilityIndexBuilder.create(
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/core/traceability_index_builder.py", line 145, in create
    traceability_index.create_traceability_info(
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/core/traceability_index.py", line 378, in create_traceability_info
    self._file_traceability_index.create_traceability_info(
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/core/file_traceability_index.py", line 444, in create_traceability_info
    function_.name
AssertionError: main(void)

```

## Environment

 - OS: Ubuntu Linux 24.04
 - Version: strictdoc-0.5.0

## Additional context

> The current behavior will be that the definitions/implementations with the same name will all inherit the requirements defined in the declaration/header. Is this what we want?

Yes, I believe this is what I would like at least.

0.6.1a1 fails like below for me (0.6.0 works):

```
Published: SW Test Specification - Control System ........... 0.03s
Published: SW Test Specification - Safety Instrumented System . 0.02s
Published: Software Verification Report - Control System .... 0.03s
Published: Vehicle Technical Requirements ................... 0.15s
Traceback (most recent call last):
  File "/home/johan/virtual_python/bin/strictdoc", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 194, in main
    raise exc
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 189, in main
    _main(parallelizer)
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/cli/main.py", line 75, in _main
    export_action.export()
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/helpers/timing.py", line 18, in wrap
    result = func(*args, **kw)
             ^^^^^^^^^^^^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/core/actions/export_action.py", line 87, in export
    html_generator.export_complete_tree(
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/export/html/html_generator.py", line 113, in export_complete_tree
    self.export_project_statistics(traceability_index)
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/export/html/html_generator.py", line 550, in export_project_statistics
    document_content = ProgressStatisticsGenerator.export(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/export/html/generators/project_statistics.py", line 42, in export
    if not node.has_any_text_nodes():
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/backend/sdoc/models/section.py", line 134, in has_any_text_nodes
    return any(
           ^^^^
  File "/home/johan/virtual_python/lib/python3.12/site-packages/strictdoc/backend/sdoc/models/section.py", line 135, in <genexpr>
    not node_.is_section and node_.node_type == "TEXT"
        ^^^^^^^^^^^^^^^^
AttributeError: 'DocumentFromFile' object has no attribute 'is_section'
make[3]: *** [reqmgmt/docs/CMakeFiles/strictdoc.dir/build.make:73: reqmgmt/docs/CMakeFiles/strictdoc] Error 1
make[2]: *** [CMakeFiles/Makefile2:1742: reqmgmt/docs/CMakeFiles/strictdoc.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1749: reqmgmt/docs/CMakeFiles/strictdoc.dir/rule] Error 2
make: *** [Makefile:738: strictdoc] Error 2

```

--- a/strictdoc/__init__.py
+++ b/strictdoc/__init__.py
@@ -1,6 +1,6 @@
 from strictdoc.core.environment import SDocRuntimeEnvironment
 
-__version__ = "0.6.0"
+__version__ = "0.6.1a2"
 
 
 environment = SDocRuntimeEnvironment(__file__)

--- a/strictdoc/backend/sdoc/models/section.py
+++ b/strictdoc/backend/sdoc/models/section.py
@@ -132,7 +132,7 @@ class SDocSection(SDocObject):
 
     def has_any_text_nodes(self):
         return any(
-            not node_.is_section and node_.node_type == "TEXT"
+            node_.__class__.__name__ == "SDocNode" and node_.node_type == "TEXT"
             for node_ in self.section_contents
         )
 

--- a/strictdoc/core/file_traceability_index.py
+++ b/strictdoc/core/file_traceability_index.py
@@ -50,7 +50,7 @@ class FileTraceabilityIndex:
 
         # This is only public non-static functions from languages like C.
         self.map_all_function_names_to_definition_functions: Dict[
-            str, Function
+            str, List[Function]
         ] = {}
 
         # "file.py" -> (
@@ -241,70 +241,77 @@ class FileTraceabilityIndex:
                     and function_.name
                     in self.map_all_function_names_to_definition_functions
                 ):
-                    definition_function: Optional[Function]
+                    definition_functions: List[Function] = []
                     if not function_.is_public():
                         definition_function = traceability_info_.ng_map_names_to_definition_functions.get(
-                            function_.name
+                            function_.name, None
                         )
+                        if definition_function is not None:
+                            definition_functions.append(definition_function)
                     else:
-                        definition_function = (
+                        mapped_definition_functions = (
                             self.map_all_function_names_to_definition_functions[
                                 function_.name
                             ]
                         )
-                    if definition_function is None:
+                        definition_functions.extend(mapped_definition_functions)
+                    if len(definition_functions) == 0:
                         continue
-                    definition_function_trace_info: SourceFileTraceabilityInfo = definition_function.parent
 
-                    for marker_ in function_.markers:
-                        function_marker = ForwardFunctionRangeMarker(
-                            parent=None,
-                            reqs_objs=marker_.reqs_objs,
-                            scope="function",
-                        )
-                        function_marker.ng_source_line_begin = (
-                            definition_function.line_begin
-                        )
-                        function_marker.ng_source_column_begin = 1
-                        function_marker.ng_range_line_begin = (
-                            definition_function.line_begin
-                        )
-                        function_marker.ng_range_line_end = (
-                            definition_function.line_end
-                        )
-                        function_marker.ng_marker_line = (
-                            definition_function.line_begin
-                        )
-                        function_marker.ng_marker_column = 1
-                        function_marker.set_description(
-                            f"function {function_.display_name}"
-                        )
+                    for definition_function_ in definition_functions:
+                        definition_function_trace_info: SourceFileTraceabilityInfo = definition_function_.parent
 
-                        for req_uid_ in marker_.reqs:
-                            markers = definition_function_trace_info.ng_map_reqs_to_markers.setdefault(
-                                req_uid_, []
+                        for marker_ in function_.markers:
+                            function_marker = ForwardFunctionRangeMarker(
+                                parent=None,
+                                reqs_objs=marker_.reqs_objs,
+                                scope="function",
                             )
-                            markers.append(function_marker)
-
-                            path_to_info = reversed_trace_info[
-                                definition_function_trace_info
-                            ]
-                            self.map_reqs_uids_to_paths.setdefault(
-                                req_uid_, OrderedSet()
+                            function_marker.ng_source_line_begin = (
+                                definition_function_.line_begin
                             )
-                            self.map_reqs_uids_to_paths[req_uid_].add(
-                                path_to_info
+                            function_marker.ng_source_column_begin = 1
+                            function_marker.ng_range_line_begin = (
+                                definition_function_.line_begin
+                            )
+                            function_marker.ng_range_line_end = (
+                                definition_function_.line_end
+                            )
+                            function_marker.ng_marker_line = (
+                                definition_function_.line_begin
+                            )
+                            function_marker.ng_marker_column = 1
+                            function_marker.set_description(
+                                f"function {function_.display_name}"
                             )
 
-                            node = traceability_index.get_node_by_uid(req_uid_)
-                            self.map_paths_to_reqs.setdefault(
-                                path_to_info, OrderedSet()
-                            )
-                            self.map_paths_to_reqs[path_to_info].add(node)
+                            for req_uid_ in marker_.reqs:
+                                markers = definition_function_trace_info.ng_map_reqs_to_markers.setdefault(
+                                    req_uid_, []
+                                )
+                                markers.append(function_marker)
 
-                        definition_function_trace_info.markers.append(
-                            function_marker
-                        )
+                                path_to_info = reversed_trace_info[
+                                    definition_function_trace_info
+                                ]
+                                self.map_reqs_uids_to_paths.setdefault(
+                                    req_uid_, OrderedSet()
+                                )
+                                self.map_reqs_uids_to_paths[req_uid_].add(
+                                    path_to_info
+                                )
+
+                                node = traceability_index.get_node_by_uid(
+                                    req_uid_
+                                )
+                                self.map_paths_to_reqs.setdefault(
+                                    path_to_info, OrderedSet()
+                                )
+                                self.map_paths_to_reqs[path_to_info].add(node)
+
+                            definition_function_trace_info.markers.append(
+                                function_marker
+                            )
 
         for (
             traceability_info_
@@ -440,14 +447,13 @@ class FileTraceabilityIndex:
             marker_type: RangeMarkerType
 
             if FunctionAttribute.DEFINITION in function_.attributes:
-                assert (
-                    function_.name
-                    not in self.map_all_function_names_to_definition_functions
-                ), function_.name
                 if function_.is_public():
+                    self.map_all_function_names_to_definition_functions.setdefault(
+                        function_.name, []
+                    )
                     self.map_all_function_names_to_definition_functions[
                         function_.name
-                    ] = function_
+                    ].append(function_)
 
             if (
                 source_file.in_doctree_source_file_rel_path_posix

